docs/development.md

```diff
@@ -0,0 +1,775 @@
+          async executionHandler (
+            executionContext: IntegrationStepExecutionContext,
+          ) {
+            return fetchUsers(executionContext);
+          },
+        },
+        {
+          id: 'fetch-groups',
+          name: 'Fetch Groups',
+          types: ['my_integration_group'],
+          executionHandler (
+            executionContext: IntegrationStepExecutionContext,
+          ) {
+            return fetchGroups(executionContext);
+          },
+        },
+      ],
+    },
+    {
+      id: 'build-relationships',
+      name: 'Collect accounts',
+      steps: [
+        {
+          id: 'build-user-to-group-relationships',
+          name: 'Fetch Accounts',
+          types: ['my_integration_user_to_group_relationship'],
+          dependsOn: ['fetch-users', 'fetch-groups'], // alternatively ['users-and-groups-phase']
+          async executionHandler (
+            executionContext: IntegrationStepExecutionContext,
+          ) {
+            return fetchAccounts(executionContext);
+          },
+        },
+      ],
+    },
+
+  ],
+};
+```
+
+### `IntegrationInvocationConfig` fields
+
+#### `instanceConfigFields`
+
+The `instanceConfigFields` field contains a map of integration fields that are
+required to be provided for authenticating with a given provider API. This
+varies between services.
+
+#### `invocationValidator`
+
+The `invocationValidator` field is a validation function that is required for
+ensuring the integration has received a valid set of `instanceConfigFields`. It
+is common practice to execute a test API call to ensure everything has been
+configured properly.
+
+#### `getStepStartStates`
+
+The `getStepStartStates` is an optional function that can be provided for
+determining if a certain phase or step should be run or not.
+
+#### `integrationStepPhases`
+
+The `integrationStepPhases` field is used to define how an integration collects
+data. It is expected that an array of phases that contain "steps" to perform the
+collection of data needs to be passed in. Phases are run in sequence and steps
+are run in parallel. This allows for data from a previous phase to be referenced
+in a later stage.
+
+A `phase` must contain an `id`, `name`, and array of `steps` to perform.
+
+A `step` must contain an `id`, `name`, list of `types` that the step expects to
+generate and an `executionHandler` function that performs the data collection
+work. It is important to provide `types` because that data will be used to
+provide context for the backend synchronization process and determine how
+updates and deletes should be applied.
+
+Optionally, a `step` can contain a `dependsOn` list that references previous
+phases and steps. This field will be used to determine if previous work has
+failed to complete and if the synchronization process should treat the data
+retrieved in the step as a partial dataset. See the [Failure handling](#Failure
+handling) section below for more information.
```

Comment:
If I have five independent resources, why wouldn't I configure them all to run as steps? Is there ever a time that I'd want to process a resource in a new phase if it _didn't_ somehow depend on the resources processed in the preceding phases? In other words, is a `dependsOn` relationship with the previous phases implied simply by processing my resource in a new phase?

Reply:
🤔 Great question. I'm not sure a dependency is always implied. There have been times when I wanted to group steps so they could fetch their data concurrently, but they had no relationship to each other. A later phase would fetch other unrelated things where some of those steps could have dependencies on one of the previous phase steps, but it would not be possible for the computer to know which steps were dependent.

Reply:
@aiwilliams In your example, you have phase 1 and 2, where some of the steps in phase 2 are dependent on the results of steps in phase 1 but some are not. Is there a reason you didn't move the independent steps in phase 2 into phase 1?

But I do hear what you're saying about having steps in a later phase depend on only some of the steps in an earlier phase such that implying a dependency on all steps in that earlier phase would be unnecessary.

Reply:
Yeah, if you have five independent resources that do not rely on each other, you would probably just want to configure them as just steps. If there was a resource that did not rely on previously collected resources, you would never really want to have it be processed in a later step since wouldn't make much sense waiting to collect that data.

And you make a good point here, the `dependsOn` relationship is seemingly implied by the previous phases considering the way they define an order of operations for the integration run. However (with the current setup), I don't think that the implied relationship there should be used to determine partial datasets. Mainly because the structure of phases and steps is too limited to be able to create an implicit dependency graph that is capable of determining partial datasets without causing later phases + steps to be marked as partial. For example, a resource collected in one phase could fail to be retrieved and cause data in a later phase that could be unrelated to be marked as partial.

I do think it's important to be able to specify the prerequisites that need to be fulfilled to determine partial datasets. I'm curious if we should potentially consider going with a different structure to support building integrations with a structure that allows for integrations to be built with implicit dependencies. Another potential option there would be to maybe do away with phases entirely and consider going with a more terraform like approach with explicit dependencies and have our state machine construct a graph of work to perform and manage the execution based on that.

Reply:
@charlieduong94 Makes sense; also, sounds good. I'm in favor of copying Hashicorp wherever possible. :laughing:

Reply:
> Another potential option there would be to maybe do away with phases entirely and consider going with a more terraform like approach with explicit dependencies and have our state machine construct a graph of work to perform and manage the execution based on that.

Yah, I like this idea very much. It makes a lot of sense to be explicit in dependencies, and would get rid of the phase/step thing and just allow developers to write data fetching work code, with dependencies declared, and then also relationship building work that depends on all fetching work to complete.